Separate staged status from GetDiff

gerhardol opened the pull request:
This was done as a part of #7902 but separated

## Proposed changes

Separate the evaluation of getting "staged status" to a separate function and add tests.
The Git documentation mostly uses stage/unstage as the action to move between the indexes, this text uses staged/unstaged to some extent too.

git-status reports if a file is changed in the worktree or index. This is used in FormCommit.
RevisionDiff allows comparing any commit to worktree/index and uses git-diff.
RevDiff allows the user to stage/unstage files if conditions allow.

The problem here is that git-diff reports the differences for the file but not if a file is changed in any of worktree, index or a intermediate commit. Therefore the staged status is calculated separately from commit. So if a file is changed from index to worktree, it must be changed in the index and it can be staged. If a commit other than index is compared to worktree, the staged status is Unknown. If two normal commits are compared, the status is None.

The 'Staged' calculation was bundled with the parsing of git-diff, it was a little confusing why "parent to second" had to be provided in the API, staged status is therefore evaluated explicitly (or provided manually.)

This PR also changes GetSingleDiff() from using string to ObjectId to be able to use ObjectId in the broken out method.
This requires that FormStash will have to do git-rev-parse. So while slightly more efficient in most forms, 40*2 milliseconds extra to open FormStash for me.  

The refactoring also corrected that status was set to None in some situations it should have been set to Unknown, and that the default should be Unset so that was added. The problem does not make a difference as only explicit WorkTree and Index is used, but the assert in GitItemStatus had to be removed.

It could be possible to set staged status only when parsing git-status, and let git-diff dynamically calculate the staged status from the compared commits. This requires a GitItemStatusWithParents as added in #7899 
(I do not propose that this is changed).

If you have read all the way here: This is not an important change, but some of the lengthy explanation will not be needed after the change.

## Test methodology 
Tests added

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented:
Thank you for the detailed write up 👍.
I think it is worth while to consolidate the code and use a consistent
approach to solving the problems.
Git allows hundreds different ways to achieve the same thing and shoot
yourself in a foot while doing that. It'd be good to pick one and stick
with it.

On Mon, Mar 30, 2020, 4:58 AM Gerhard Olsson <notifications@github.com>
wrote:

> This was done as a part of #7902
> <https://github.com/gitextensions/gitextensions/issues/7902> but separated
> Proposed changes
>
> Separate the evaluation of getting "staged status" to a separate function
> and add tests.
> The Git documentation mostly uses stage/unstage as the action to move
> between the indexes, this text uses staged/unstaged to some extent too.
>
> git-status reports if a file is changed in the worktree or index. This is
> used in FormCommit.
> RevisionDiff allows comparing any commit to worktree/index and uses
> git-diff.
> RevDiff allows the user to stage/unstage files if conditions allow.
>
> The problem here is that git-diff reports the differences for the file but
> not if a file is changed in any of worktree, index or a intermediate
> commit. Therefore the staged status is calculated separately from commit.
> So if a file is changed from index to worktree, it must be changed in the
> index and it can be staged. If a commit other than index is compared to
> worktree, the staged status is Unknown. If two normal commits are compared,
> the status is None.
>
> The 'Staged' calculation was bundled with the parsing of git-diff, it was
> a little confusing why "parent to second" had to be provided in the API,
> staged status is therefore evaluated explicitly (or provided manually.)
>
> This PR also changes GetSingleDiff() from using string to ObjectId to be
> able to use ObjectId in the broken out method.
> This requires that FormStash will have to do git-rev-parse. So while
> slightly more efficient in most forms, 40*2 milliseconds extra to open
> FormStash for me.
>
> The refactoring also corrected that status was set to None in some
> situations it should have been set to Unknown, and that the default should
> be Unset so that was added. The problem does not make a difference as only
> explicit WorkTree and Index is used, but the assert in GitItemStatus had to
> be removed.
>
> It could be possible to set staged status only when parsing git-status,
> and let git-diff dynamically calculate the staged status from the compared
> commits. This requires a GitItemStatusWithParents as added in #7899
> <https://github.com/gitextensions/gitextensions/pull/7899>
> (I do not propose that this is changed).
>
> If you have read all the way here: This is not an important change, but
> some of the lengthy explanation will not be needed after the change.
> Test methodology
>
> Tests added
> ------------------------------
>
> ✒️ I contribute this code under The Developer Certificate of Origin
> <http://../blob/master/contributors.txt>.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/gitextensions/gitextensions/pull/7914
> Commit Summary
>
>    - Separate staged status from GetDiff
>
> File Changes
>
>    - *M* GitCommands/Git/GitCommandHelpers.cs
>    <https://github.com/gitextensions/gitextensions/pull/7914/files#diff-04ee9454ae88464d76063370410bd2fa>
>    (35)
>    - *M* GitCommands/Git/GitItemStatus.cs
>    <https://github.com/gitextensions/gitextensions/pull/7914/files#diff-5f321c1f15fb5e115fe8a344cdbf4a98>
>    (19)
>    - *M* GitCommands/Git/GitModule.cs
>    <https://github.com/gitextensions/gitextensions/pull/7914/files#diff-05956b8e9b35344894a0ffe609cf32b9>
>    (23)
>    - *M* GitUI/CommandsDialogs/FormArchive.cs
>    <https://github.com/gitextensions/gitextensions/pull/7914/files#diff-12e660db24ac6bbd17ab34df1b9671b6>
>    (3)
>    - *M* GitUI/CommandsDialogs/FormStash.cs
>    <https://github.com/gitextensions/gitextensions/pull/7914/files#diff-bd65142524329e738dd037bc350a5541>
>    (6)
>    - *M* GitUI/GitUIExtensions.cs
>    <https://github.com/gitextensions/gitextensions/pull/7914/files#diff-95349e407fef1a99246ba82fa6aa7ba0>
>    (8)
>    - *M* UnitTests/GitCommands.Tests/Git/GitCommandHelpersTest.cs
>    <https://github.com/gitextensions/gitextensions/pull/7914/files#diff-393a87f649e4282850e7d1c0eac2b126>
>    (70)
>
> Patch Links:
>
>    - https://github.com/gitextensions/gitextensions/pull/7914.patch
>    - https://github.com/gitextensions/gitextensions/pull/7914.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/7914>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXVXQXN5PKJF3JUSCA3RJ6D5DANCNFSM4LWB5O7A>
> .
>

mstv commented on GitUI/CommandsDialogs/FormArchive.cs:
```diff
@@ -160,7 +160,8 @@ private string GetPathArgumentFromGui()
             {
                 // 1. get all changed (and not deleted files) from selected to current revision
-                var files = UICommands.Module.GetDiffFiles(DiffSelectedRevision.Guid, SelectedRevision.Guid, SelectedRevision.ParentIds.FirstOrDefault()?.ToString()).Where(f => !f.IsDeleted);
+                var stagedStatus = GitCommandHelpers.GetStagedStatus(DiffSelectedRevision?.ObjectId, SelectedRevision?.ObjectId, DiffSelectedRevision?.ObjectId);
```
In a pure refactoring, I had expected:
```suggestion
                var stagedStatus = GitCommandHelpers.GetStagedStatus(DiffSelectedRevision?.ObjectId, SelectedRevision?.ObjectId, SelectedRevision?.ParentIds?.FirstOrDefault());
```

gerhardol replied:
I believe I had doubt if parents were set for the revision, but DiffSelectedRevision is not always the parent so this could be incorrect
However, comparisions cannot be done to artificial, so StagedStatus is always None

Rebasing and squashing this

RussKie commented:
I don't have a good understanding of the functionality, so can't comment on it.
From the code stand point - the change looks ok. 

I have pushed a refactor to the test.

gerhardol commented:
Squashed, refreshed some comments and added a testcase, rebased on #7938 (conflicts)
To be rebased again and merged within a day when #7638 is merged
